hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/JDBCSource.java

```diff
@@ -0,0 +1,233 @@
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SparkSession;
+import org.apache.spark.sql.functions;
+import org.apache.spark.sql.types.DataTypes;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class JDBCSource extends RowSource {
+
+  private static Logger LOG = LoggerFactory.getLogger(JDBCSource.class);
+
+  private final String ppdQuery = "(select * from %s where %s >= \" %s \") rdbms_table";
+
+
+  public JDBCSource(TypedProperties props, JavaSparkContext sparkContext, SparkSession sparkSession,
+      SchemaProvider schemaProvider) {
+    super(props, sparkContext, sparkSession, schemaProvider);
+  }
+
+  private static DataFrameReader validatePropsAndGetDataFrameReader(final SparkSession session,
+      final TypedProperties properties)
+      throws HoodieException {
+    FSDataInputStream passwordFileStream = null;
+    try {
+      DataFrameReader dataFrameReader = session.read().format("jdbc");
+      dataFrameReader = dataFrameReader.option(Config.URL_PROP, properties.getString(Config.URL));
+      dataFrameReader = dataFrameReader.option(Config.USER_PROP, properties.getString(Config.USER));
+      dataFrameReader = dataFrameReader.option(Config.DRIVER_PROP, properties.getString(Config.DRIVER_CLASS));
+      dataFrameReader = dataFrameReader
+          .option(Config.RDBMS_TABLE_PROP, properties.getString(Config.RDBMS_TABLE_NAME));
+
+      if (!properties.containsKey(Config.PASSWORD)) {
+        if (properties.containsKey(Config.PASSWORD_FILE)) {
+          if (!StringUtils.isNullOrEmpty(properties.getString(Config.PASSWORD_FILE))) {
+            LOG.info("Reading password for password file {}", properties.getString(Config.PASSWORD_FILE));
+            FileSystem fileSystem = FileSystem.get(new Configuration());
+            passwordFileStream = fileSystem.open(new Path(properties.getString(Config.PASSWORD_FILE)));
+            byte[] bytes = new byte[passwordFileStream.available()];
+            passwordFileStream.read(bytes);
+            dataFrameReader = dataFrameReader.option(Config.PASSWORD_PROP, new String(bytes));
+          } else {
+            throw new IllegalArgumentException(
+                String.format("%s property cannot be null or empty", Config.PASSWORD_FILE));
+          }
+        } else {
+          throw new IllegalArgumentException(String.format("JDBCSource needs either a %s or %s to connect to RDBMS "
+              + "datasource", Config.PASSWORD_FILE, Config.PASSWORD));
+        }
+      } else if (!StringUtils.isNullOrEmpty(properties.getString(Config.PASSWORD))) {
+        dataFrameReader = dataFrameReader.option(Config.PASSWORD_PROP, properties.getString(Config.PASSWORD));
+      } else {
+        throw new IllegalArgumentException(String.format("%s cannot be null or empty. ", Config.PASSWORD));
+      }
+      if (properties.containsKey(Config.EXTRA_OPTIONS)) {
```

Comment:
instead of splitting/ecnoding using `,` like this.. can you just take all properites that begin with the prefix. `hoodie.datasource.jdbc`

Reply:
The extra.options is for adding some properties for jdbc pulls. so lets say use writes the below spark code.
`val df1 = spark.read.format("jdbc")
      .option("url", jdbcUrl)
      .option("dbtable", "datetimePartitionTest")
      .option("partitionColumn", "d")
      .option("lowerBound", "2018-07-06")
      .option("upperBound", "2018-07-20")
      .option("numPartitions", 3)
      // oracle.jdbc.mapDateToTimestamp defaults to true. If this flag is not disabled, column d
      // (Oracle DATE) will be resolved as Catalyst Timestamp, which will fail bound evaluation of
      // the partition column. E.g. 2018-07-06 cannot be evaluated as Timestamp, and the error
      // message says: Timestamp format must be yyyy-mm-dd hh:mm:ss[.fffffffff].
      .option("oracle.jdbc.mapDateToTimestamp", "false")
      .option("sessionInitStatement", "ALTER SESSION SET NLS_DATE_FORMAT = 'YYYY-MM-DD'")
      .load()`
 Now how do we let the user express the same with a properties file? We use extra.options

Reply:
> The extra.options is for adding some properties for jdbc pulls. so lets say use writes the below spark code.
> 
> `val df1 = spark.read.format("jdbc") .option("url", jdbcUrl) .option("dbtable", "datetimePartitionTest") .option("partitionColumn", "d") .option("lowerBound", "2018-07-06") .option("upperBound", "2018-07-20") .option("numPartitions", 3) .option("timestampFormat", "MM/dd/yyyy h:mm:ss a") .option("oracle.jdbc.mapDateToTimestamp", "false") .option("sessionInitStatement", "ALTER SESSION SET NLS_DATE_FORMAT = 'YYYY-MM-DD'") .load()`
> 
> Now how do we let the user express the same with a properties file? We use extra.options
> 
> `hoodie.datasource.jdbc.extra.options=" fetchsize=1000, timestampFormat=MM/dd/yyyy h:mm:ss a, oracle.jdbc.mapDateToTimestamp=false, sessionInitStatement="ALTER SESSION SET NLS_DATE_FORMAT = 'YYYY-MM-DD'"`

I'm seeing users using "=" in some of the options they pass. Hence I think we should reconsider what the delimiter should be. I think `||` should do it. Thinking of how no one ever uses a double pipe.

Reply:
Or simply do 
```
hoodie.deltastreamer.source.jdbc.fetchsize=1000
hoodie.deltastreamer.source.jdbc.timestampFormat=...
```

and so on? In your code, just look for properties with prefix `hoodie.deltastreamer.source..` . I think KafkaSource works like this..  You can check it out..